writeengine/bulk/we_bulkloadbuffer.cpp

```diff
@@ -515,14 +513,32 @@ void BulkLoadBuffer::convert(char* field, int fieldLength, bool nullFlag, unsign
         // from storing characters beyond the column's defined width.
         // It contains the column definition width rather than the bytes
         // on disk (e.g. 5 for a varchar(5) instead of 8).
-        if (fieldLength > column.definedWidth)
+        if (column.cs->mbmaxlen > 1)
         {
-          uint8_t truncate_point = utf8::utf8_truncate_point(field, column.definedWidth);
-          memcpy(charTmpBuf, field, column.definedWidth - truncate_point);
-          bufStats.satCount++;
+          const CHARSET_INFO* cs = column.cs;
+          const char* start = (const char*) field;
+          const char* end = (const char*)(field + fieldLength);
+          size_t numChars = cs->numchars(start, end);
```

Comment:
This op might have `O(N)` complexity where `N` is a length of utf8 string. AFAIK numchars() will run full blown decoding that is complex 3 tables(in the worst-case) lookup. What I suggest is to set a length using maxCharLength and skip numChars calcultion. Surely if this is possible.

Reply:
Yes, `cs->numchars()` will be `O(N)` where N is the number of bytes in the input string.

For `utf8mb3` charset (as an example), here is the flow of execution for `cs->numchars()`:
```c++
my_numchars_mb() (defined in string/ctype-mb.c) -> my_ismbchar() (defined in include/m_ctype.h) -> my_valid_mbcharlen_utf8mb3() (defined in strings/ctype-utf8.c)
```
In this execution flow, I don't see a 3-table complex lookup anywhere.

I don't see how it is possible to determine if a string needs to be truncated or not without first finding it's character length.